File: platform-nest/src/events/hlc.test.ts
import { describe, it, expect } from "vitest";
import { HlcClock, formatHlc, parseHlc } from "./hlc";

describe("HlcClock", () => {
  it("is monotonic under identical wall time (counter bumps)", () => {
    const c = new HlcClock(() => 1000);
    const h1 = c.next();
    const h2 = c.next();
    expect(h2 > h1).toBe(true); // padded text ordering == logical ordering
  });

  it("resets the counter when wall time advances", () => {
    let wall = 1000;
    const c = new HlcClock(() => wall);
    const h1 = c.next();
    wall = 2000;
    const h2 = c.next();
    expect(parseHlc(h2)).toEqual({ wallMs: 2000, counter: 0 });
    expect(h2 > h1).toBe(true);
  });

  it("seedFromPersisted never lets the clock regress below the last known HLC", () => {
    const c = new HlcClock(() => 500); // simulate a regressed/skewed wall clock on restart
    c.seedFromPersisted(formatHlc(9000, 3));
    const h = c.next();
    expect(h > formatHlc(9000, 3)).toBe(true);
  });

  it("padded format keeps text ordering equal to logical ordering across digit widths", () => {
    // The bug this guards: unpadded "1000.0" < "999.0" lexicographically. Padded must not.
    expect(formatHlc(1000, 0) > formatHlc(999, 0)).toBe(true);
    expect(formatHlc(1000, 10) > formatHlc(1000, 9)).toBe(true);
  });

  it("round-trips through parseHlc", () => {
    expect(parseHlc(formatHlc(123456789, 7))).toEqual({ wallMs: 123456789, counter: 7 });
  });
});

File: sync-engine-go/internal/hlc/hlc_test.go
package hlc

import "testing"

func TestClockIsMonotonicUnderSameWallTime(t *testing.T) {
	wall := int64(1000)
	c := NewClock(func() int64 { return wall })
	h1 := c.Next()
	h2 := c.Next()
	if h2.Compare(h1) <= 0 {
		t.Fatalf("expected h2 > h1 even under identical wall time, got h1=%v h2=%v", h1, h2)
	}
}

func TestClockAdvancesWithWallTime(t *testing.T) {
	wall := int64(1000)
	c := NewClock(func() int64 { return wall })
	h1 := c.Next()
	wall = 2000
	h2 := c.Next()
	if h2.WallMs != 2000 || h2.Counter != 0 {
		t.Fatalf("expected wall-time jump to reset counter, got %+v", h2)
	}
	if h2.Compare(h1) <= 0 {
		t.Fatal("expected h2 > h1")
	}
}

func TestSeedFromPersistedRejectsRegression(t *testing.T) {
	wall := int64(500) // simulates a clock-skewed/regressed node on failover
	c := NewClock(func() int64 { return wall })
	c.SeedFromPersisted(HLC{WallMs: 9000, Counter: 3})
	h := c.Next()
	if h.Compare(HLC{WallMs: 9000, Counter: 3}) <= 0 {
		t.Fatalf("expected seeded clock never to regress below the last persisted HLC, got %+v", h)
	}
}

func TestStringIsPaddedSoTextOrderMatchesLogicalOrder(t *testing.T) {
	// The bug this guards: unpadded "1000.0" < "999.0" lexicographically.
	if !(HLC{WallMs: 1000}.String() > HLC{WallMs: 999}.String()) {
		t.Fatal("padded text ordering must match logical ordering across digit widths")
	}
	if !(HLC{WallMs: 1000, Counter: 10}.String() > HLC{WallMs: 1000, Counter: 9}.String()) {
		t.Fatal("counter must be padded too")
	}
}

func TestParseRoundTrips(t *testing.T) {
	h := HLC{WallMs: 123456789, Counter: 7}
	parsed, err := Parse(h.String())
	if err != nil {
		t.Fatalf("Parse failed: %v", err)
	}
	if parsed != h {
		t.Fatalf("round-trip mismatch: %+v != %+v", parsed, h)
	}
}

// Interop guard: the Go padded form must equal what platform-nest's formatHlc emits.
func TestPaddedFormatMatchesTypeScriptContract(t *testing.T) {
	if got := (HLC{WallMs: 1700000000000, Counter: 0}).String(); got != "1700000000000.0000000000" {
		t.Fatalf("format drift vs platform-nest hlc.ts: got %q", got)
	}
}
